<launch>
  
  <!-- Load hand geometry parameters -->  
  <include file="$(find gpd)/launch/hand_geometry.launch">
    <arg name="node" value="classify_grasp_candidates" />
  </include>
  
  <!-- Load classifier parameters -->  
  <include file="$(find gpd)/launch/caffe/ur5_classifier_15channels.launch">
    <arg name="node" value="classify_grasp_candidates" />
  </include>
  
	<node name="classify_grasp_candidates" pkg="gpd" type="classify_candidates" output="screen">
    
    <!-- Where the point cloud is stored -->    
<!-- <param name="cloud_file_name" value="$(find gpd)/tutorials/table_mug.pcd" /> -->
<!-- <param name="cloud_file_name" value="/media/andreas/storage/data/bigbird_pcds/white_rain_sensations_apple_blossom_hydrating_body_wash_gt.pcd" /> -->
    <param name="cloud_file_name" value="/media/andreas/storage/data/baxter_clutter/scene0.pcd" />

    <!-- If sequential importance sampling is used -->
    <param name="use_importance_sampling" value="true" />
    
    <!-- Parameters for sequential importance sampling -->
    <param name="num_init_samples" value="50" />
    <param name="num_iterations" value="10" />    
    <param name="num_samples_per_iteration" value="20" />
    <param name="prob_rand_samples" value="0.2" />
    <param name="std" value="0.01" />
    <param name="sampling_method" value="1" /> <!-- 0: Sum of Gaussians, 1: Max of Gaussians -->
    
    <param name="visualize_rounds" value="false" />
    <param name="visualize_steps" value="false" />
    <param name="visualize_results" value="true" />
            
    <!-- Preprocessing of point cloud -->
    <param name="voxelize" value="true"/> <!-- Voxelize the point cloud -->
    <param name="remove_outliers" value="false"/> <!-- Remove statistical outliers from the point cloud -->
<!-- <rosparam param="workspace"> [-1, 1, -1, 1, -1, 1] </rosparam> --> <!-- (x,y,z)-dimensions of the workspace --> 
    <rosparam param="workspace"> [0.5, 0.9, -0.4, 0.25, -0.28, 1] </rosparam> <!-- (x,y,z)-dimensions of the workspace -->
    <rosparam param="camera_position"> [0, 0, 0] </rosparam> <!-- The position of the camera -->

    <!-- General parameters -->
    <param name="num_threads" value="4" /> <!-- The number of CPU threads to be used -->
    
    <!-- Parameters for local grasp candidate search -->
    <param name="nn_radius" value="0.01" /> <!-- Radius for the point neighborhood search -->
    <param name="num_orientations" value="8" /> <!-- Number of orientations to consider -->
    
    <!-- Filtering of grasp candidates --> 
    <param name="filter_grasps" value="false" /> <!-- on workspace and robot gripper width -->
    <param name="min_aperture" value="0.3" /> <!-- The minimum gripper width -->
    <param name="max_aperture" value="0.7" /> <!-- The maximum gripper width -->
    <rosparam param="workspace_grasps"> [0.57, 0.87, -0.38, 0.23, -0.26, 1] </rosparam> <!-- (x,y,z)-dimensions of the workspace, should be smaller than <workspace> -->
    <param name="filter_half_antipodal" value="true"/> <!-- on half antipodal -->
    
    <!-- Filtering of grasps which are too low on (i.e. too close to) table -->
    <param name="filter_table_side_grasps" value="false" />
    <rosparam param="vertical_axis"> [0, 0, 1] </rosparam>
    <param name="angle_thresh" value="0.1" />
    <param name="table_height" value="0.0" />
    <param name="table_thresh" value="0.05" />
    
    <!-- Grasp image creation -->
    <param name="create_image_batches" value="false" /> <!-- creates grasp images in batches (less memory usage) -->
        
    <!-- Clustering of grasps -->
    <param name="min_inliers" value="0" /> <!-- min inliers in each cluster; set to 0 to avoid clustering -->
    
    <!-- Grasp selection -->
    <param name="min_score_diff" value="50" /> <!-- Threshold to sort out grasps with low scores -->
    <param name="num_selected" value="100" /> <!-- The number of selected grasps (sorted by score) -->
        
	</node>
  
</launch>
